### ISSUE TITLE
The module.update_metric may hurt the performance in some cases.

### ISSUE THREAD

**Issue by huyangc** [apache/incubator-mxnet#11089]
I found that the input of the metric instance(preds and labels) might be in different context, the preds will always in ``cpu``. So that when I train a network that have about 100 k classes, the output of the mod will be copied to cpu to calculate the cross entropy loss. The copy procedure will hurt the performance. For example, when I train a resnet-18 with 100k classes:
Speed with CrossEntropy: ~200 fps
Speed without Cross Entropy: ~2000 fps

Why when update metric, just calculate the result in the gpu context and get the final result and then copy to cpu for showing?

**Comment by ThomasDelteil** [comment]
Have a look at this issue for this problem, https://github.com/apache/incubator-mxnet/issues/9571

If you use Gluon, you can compute easily your metric directly on GPU and indeed you can get a lot of performance gain.

However be careful on your evaluation, if you remove all the blocking calls (asnumpy(), metric calculation, loss printing, etc) you could have the false impression that your training is completed whilst it is actually running in the background. So before completing your timing experiment, add a `mx.nd.waitall()` to force wait for all the asynchronous operation to complete.

**Comment by huyangc** [comment]
Thanks. Maybe I am wrong of the memory copy time consuming. Because the 100k floats are only 0.4M, Maybe calculating on the cpu hurt the performance.
